src/runtime/contrib/tensorrt/tensorrt_builder.cc

```diff
@@ -0,0 +1,421 @@
+  for (size_t i = 0; i < network_outputs.size(); ++i) {
+    CHECK(network_outputs[i].type == kTensor);
+    auto out_tensor = network_outputs[i].tensor;
+    std::string output_name = "tensorrt_output" + std::to_string(i);
+    out_tensor->setName(output_name.c_str());
+    network_output_names.push_back(output_name);
+    network_->markOutput(*out_tensor);
+    DLOG(INFO) << "Added TRT network output: " << out_tensor->getName()
+               << " -> " << output_name;
+  }
+  nvinfer1::ICudaEngine* engine = builder_->buildCudaEngine(*network_);
+  CHECK_EQ(engine->getNbBindings(),
+           network_input_map_.size() + network_outputs.size());
+  CleanUp();
+  nvinfer1::IExecutionContext* context = engine->createExecutionContext();
+  return {engine, context, network_input_map_, network_output_names};
+}
+
+nvinfer1::Weights TensorRTBuilder::GetDLTensorAsWeights(
+    DLTensor* dptr, DLDeviceType src_device) {
+  CHECK_EQ(dptr->ctx.device_type, src_device);
+  CHECK_EQ(static_cast<int>(dptr->dtype.code), kDLFloat);
+  const size_t weight_bytes = runtime::GetDataSize(*dptr);
+  nvinfer1::Weights weight{nvinfer1::DataType::kFLOAT, nullptr, 0};
+  size_t count = 1;
+  for (tvm_index_t i = 0; i < dptr->ndim; ++i) {
+    count *= dptr->shape[i];
+  }
+  CHECK_EQ(count * 4, weight_bytes);
+  weight.count = count;
+  weight.values = new float[count];
+  CHECK_EQ(
+      TVMArrayCopyToBytes(dptr, const_cast<void*>(weight.values), weight_bytes),
+      0)
+      << TVMGetLastError();
+  trt_weights_.push_back(weight);
+  return weight;
+}
+
+nvinfer1::Weights TensorRTBuilder::GetNdArrayAsWeights(
+    const runtime::NDArray& array, DLDeviceType src_device) {
+  DLTensor* dptr = const_cast<DLTensor*>(array.operator->());
+  return GetDLTensorAsWeights(dptr, src_device);
+}
+
+void TensorRTBuilder::GetInputAsWeights(const VarNode* node) {
+  const int var_node_idx = TrackVarNode(node);
+  nvinfer1::Weights weight =
+      GetDLTensorAsWeights(execution_args_[var_node_idx], kDLGPU);
+  node_output_map_[node] = {TrtOpInput(weight, GetShape(node->checked_type()))};
+}
+
+void TensorRTBuilder::GetConstantAsWeights(const ConstantNode* node) {
+  auto weight = GetNdArrayAsWeights(node->data, kDLCPU);
+  auto shape_long = node->data.Shape();
+  std::vector<int> shape(shape_long.begin(), shape_long.end());
+  node_output_map_[node] = {TrtOpInput(weight, shape)};
+}
+
+void TensorRTBuilder::GetInputAsTransposedWeights(const CallNode* transpose,
+                                                  const VarNode* node) {
+  GetInputAsWeights(node);
+  CHECK_EQ(node_output_map_[node].size(), 1);
+  const nvinfer1::Weights& original_weight = node_output_map_[node][0].weight;
+  const auto& original_shape = node_output_map_[node][0].weight_shape;
+  float* values = new float[original_weight.count];
+  // Get order and new shape.
+  const auto* attrs = transpose->attrs.as<TransposeAttrs>();
+  std::vector<int> order;
+  std::vector<int> new_shape;
```

Comment:
Looks like dyn mem alloc can be avoided for `order` and `new_shape` vectors. Can you use stack memory here?